cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,115 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <initializer_list>
+#include <utility>
+#include <vector>
+
+#include "arrow/util/variant.h"
+
+namespace arrow {
+namespace util {
+
+/// A minimal vector-like container which holds up to a single element without allocation.
+template <typename T>
+class small_vector {
+ public:
+  using value_type = T;
+  using iterator = T*;
+  using const_iterator = const T*;
+
+  small_vector() = default;
+
+  small_vector(std::initializer_list<value_type> list) {
```

Comment:
What happens if I call `small_vector({})`?

Reply:
an empty small_vector will be constructed